### PR TITLE
Use same thread composition in SyncResponseHandlerAdaptor. See #994

### PR DESCRIPTION
**Pull request by zoewangg**
## Motivation and Context
Related to #994 

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](../docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/async/SyncResponseHandlerAdapter.java [hunk 1]
```diff
@@ -76,7 +76,7 @@ public void onError(Throwable err) {
     @Override
     public CompletableFuture<T> prepare() {
         streamFuture = new CompletableFuture<>();
-        return streamFuture.thenComposeAsync(baos -> {
+        return streamFuture.thenCompose(baos -> {
```
Comment (dagnir) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/async/SyncResponseHandlerAdapter.java]: Hmm the stream future is completed by thread calling `onComplete` on the subscriber which will be owned by the async HTTP client. If we complete on the same thread we would block the HTTP client's thread while `responseHandler` runs. Should we use the future completion executor?

Comment (zoewangg) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/async/SyncResponseHandlerAdapter.java]: The `responseHandler` should run fairly fast. I wonder if it's worth delegating to the future completion executor.

Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/async/SyncResponseHandlerAdapter.java]: That would depend on the size and complexity of the response I suppose. A DDB `BatchGetItem` could return up to 16MB and could take a non-negligible time to un marshall for example (though I've never measured it).

Comment (zoewangg) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/async/SyncResponseHandlerAdapter.java]: Yeah, ec2 response unmarshalling could also take non-negligible time. I'd be interested to see which is more costly, unmarshalling large payload or transitioning task from one thread to another. Need to run some benchmarks.

Looking at page 63 of the CompletableFuture perf presentation slides https://qconsf.com/sf2017/system/files/presentation-slides/cf.pdf, seems like offloading executor would add large overhead for small quick task and small overhead for large tasks that take 100ms.

Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/async/SyncResponseHandlerAdapter.java]: Cool. I don't think this needs to be a blocker. We can go with a synchronous compose for now and offload if we need to (or provide the option); I think large responses are the exception not the norm.

Comment (millems) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/async/SyncResponseHandlerAdapter.java]: Something CPU-bound _should_ be fine to happen on the event loop threads. It's a decent bit of overhead to pass to another thread.

Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/async/SyncResponseHandlerAdapter.java]: @millems not sure. Ideally the event loop threads should only be handling I/O IMO. For example, if there was only 1 event loop thread, and multiple concurrent requests it shouldn't be blocked too long doing non I/O things for any one request, especially if data is waiting to be sent/received.

Comment (millems) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/async/SyncResponseHandlerAdapter.java]: Yeah, it's possible. Benchmarking will tell us the true story.

**Pull request merged**